Duration type is a builtin type

Akirathan opened the pull request:
### Pull Request Description

- Reimplement the `Duration` type to a built-in type.
- `Duration` is an interop type.
- Allow Enso method dispatch on `Duration` interop coming from different languages.

### Important Notes

- The older `Duration` type should now be split into new `Duration` builtin type and a `Period` type.
  - This PR does not implement `Period` type, so all the `Period`-related functionality is currently not working, e.g., `Date - Period`.
- This PR removes `Integer.milliseconds`, `Integer.seconds`, ..., `Integer.years` extension methods.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

Commit:
Add tests

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/enso-org/enso?pullRequest=3759) <br/>All committers have signed the CLA.

Commit:
Refactor Duration to a builtin type

Akirathan commented:
According to the associated issue description, we want to split the current functionality of `Duration` into `Duration` and `Period`, so I removed all the date-related functionality from `Duration`, and left only time-related functionality. All the removed date-related functionality should be moved to a new `Period` (maybe existing `Time_Period`?) builtin type in the future.

hubertp commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java:
```diff
@@ -138,6 +140,30 @@ public EnsoTimeZone zone() {
+  @Builtin.Method(name = "plus_builtin", description = "Adds a duration to this date time")
+  @TruffleBoundary
+  public EnsoDateTime plus(Object durationObject) {
+    InteropLibrary interop = InteropLibrary.getUncached();
```
You want to avoid using the uncached one. 
You can auto-generate the correct one by adding  `@Builtin.Specialize` and changing the definition to `public EnsoDateTime plus(Object durationObject, InteropLibrary interop) {`

hubertp commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java:
```diff
@@ -138,6 +140,30 @@ public EnsoTimeZone zone() {
+  @Builtin.Method(name = "minus_builtin", description = "Subtracts a duration from this date time")
+  @TruffleBoundary
+  public EnsoDateTime minus(Object durationObject) {
+    InteropLibrary interop = InteropLibrary.getUncached();
```
Same comment as for `plus`

hubertp commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDuration.java:
```diff
@@ -0,0 +1,157 @@
+        name = "new_builtin",
+        description = "Constructs a new Duration from hours, minutes, seconds, milliseconds and nanoseconds")
+    @TruffleBoundary
+    @Builtin.ReturningGuestObject
```
You don't want that since you added `EnsoDuration` to the list of accepted types. Need to update the annotation processor in `TypeWithKind` so that it is happy 

```suggestion
```

hubertp commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDuration.java:
```diff
@@ -0,0 +1,157 @@
+        name = "between_builtin",
+        description = "Construct a new Duration that is between the given start date inclusive, and end date exclusive")
+    @TruffleBoundary
+    @Builtin.ReturningGuestObject
```
```suggestion
```
same as for create and others

hubertp commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDuration.java:
```diff
@@ -0,0 +1,157 @@
+    @TruffleBoundary
+    @Builtin.ReturningGuestObject
+    public static EnsoDuration between(Object startInclusive, Object endExclusive) {
+        InteropLibrary interop = InteropLibrary.getUncached();
```
Similar comment about `interop` as for `plus` and `minus`

hubertp commented on test/Tests/src/Semantic/Java_Interop_Spec.enso:
```diff
@@ -53,4 +61,19 @@ spec =
             april1st.month.should_equal 4
             april1st.day.should_equal 1
 
+        Test.specify "Difference of Java Date and Enso date should be an Enso Duration" <|
```
I think you might want to add those to Duration_Spec, similarly as we have date-specific tests in `Date_Spec`, `Date_Time_Spec` etc.

hubertp reviewed (commented):
The main implementation looks fine, but there are few cleanups necessary

Commit:
Move all Duration-related tests to Duration_Spec.enso

Commit:
Get rid of uncached InteropLibrary and ReturningGuestObject annotation

Commit:
Fix style

Commit:
Java Duration is coerced to enso Duration

Commit:
EnsoDuration works with polyglot duration

Commit:
Refactor Time_Of_Day to use new Duration

Commit:
Merge branch 'develop' into wip/akirathan/builtin-duration-type-182987539

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/node/callable/InvokeMethodNode.java:
```diff
@@ -349,6 +356,35 @@ Stateful doConvertDateTime(
+    var ctx = Context.get(this);
+    try {
+      var duration = interop.asDuration(self);
+      var ensoDuration = new EnsoDuration(duration);
```
This is exact copy of date, time & co. handling. However in case of _array_ handling I decided to not do any wrapping, just let the `hasArrayElements` object flow thru the engine. Pavel, what do you think is better from Truffle perspective: wrap or let the original `TruffleObject` flow?

If _"let the original object flow"_, can you try it with duration and other `doConvertXYZ` methods as well?

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java:
```diff
@@ -3,16 +3,24 @@
 import org.enso.interpreter.runtime.library.dispatch.TypesLibrary;
 
-import java.time.*;
+import java.time.DateTimeException;
```
Replacing `*` imports with explicit once improves accuracy. I am all for it. However we should try to avoid changing imports just because our IDEs have explicit opinions on their particular order...

radeusgd replied:
Shouldn't we just agree on an ordering that we want and try to enforce it?

Technically we are using the Google Java Style [(according to our docs)](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md#code-formatting), which defines that [imports should be sorted in ASCII order](https://google.github.io/styleguide/javaguide.html#s3.3.3-import-ordering-and-spacing) (more or less).

JaroslavTulach replied:
I care mostly about not using `*` import. Anyway until we resolve [Java formatter doesn't work on files that use latest syntax](https://www.pivotaltracker.com/story/show/182515732) - we are probably left to _human judgment of common sense_.

radeusgd replied:
Btw. I've downloaded the formatter as a standalone JAR and downloaded Java 17 using SDKman and sometimes if I want to format a file using new syntax, I do:
```
/home/radeusgd/.sdkman/candidates/java/current/bin/java -jar ~/Downloads/google-java-format-1.15.0-all-deps.jar --replace path/to/java/file.java
```

and that works flawlessly :)

Of course it's a pity it's not automated, but it's some middle ground.

JaroslavTulach replied:
We'd have to put that into CI check to enforce the fomatting...

radeusgd replied:
I think we could implement such a check running on some default Java 19 JVM.

Still, until we do, I think we can just agree to try to keep the files formatted using this method, we will miss some, but I think that's better than not doing it at all.

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDuration.java:
```diff
@@ -0,0 +1,162 @@
+@ExportLibrary(InteropLibrary.class)
+@ExportLibrary(TypesLibrary.class)
+@Builtin(pkg = "date", name = "Duration", stdlibName = "Standard.Base.Data.Time.Duration")
+public class EnsoDuration implements TruffleObject {
```
`TruffleObject` cannot be subclassed as far as I know - shall we make these classes `final`?

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/type/Types.java:
```diff
@@ -52,6 +52,7 @@
   EnsoDateTime.class,
   EnsoTimeOfDay.class,
   EnsoTimeZone.class,
+  EnsoDuration.class
```
The ultimate goal is to get rid of all `TruffleObject` classes in this `Types` list. Technically nothing happens if we remove all date&time related ones. The important once to remove are however `Array` and `Vector` - tracked as [Array vs. ArrayOverBuffer](https://www.pivotaltracker.com/n/projects/2539304/stories/183058716) and [many builtins](https://www.pivotaltracker.com/n/projects/2539304/stories/183266964).

JaroslavTulach commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java:
```diff
@@ -138,6 +146,27 @@ public EnsoTimeZone zone() {
+  }
+
+  @Builtin.Method(name = "minus_builtin", description = "Subtracts a duration from this date time")
+  @Builtin.Specialize
```
**Request**: Changing builtins will very likely need changes to reflective files like in #3755 - please run `sbt engine-runner-native/buildNativeImage` to verify it builds properly. CCing @hubertp , @mwu-tow .

hubertp replied:
To maybe complement @JaroslavTulach's comment - once it finishes (it will take a while) you should be able to execute `./runner --run engine/runner-native/src/test/resources/Factorial.enso 6` without errors

JaroslavTulach commented on std-bits/base/src/main/java/org/enso/base/Time_Utils.java:
```diff
@@ -79,29 +90,6 @@ public static LocalDate date_adjust(LocalDate date, AdjustOp op, Period duration
   }
 
-  public static ZonedDateTime datetime_adjust(
-      ZonedDateTime datetime, AdjustOp op, Period period, Duration duration) {
```
Surprising. How do you handle `Period`? It is not Truffle interop type and needs to be still treated via host interop, I think.

JaroslavTulach commented on test/Tests/src/Data/Time/Duration_Spec.enso:
```diff
@@ -3,99 +3,88 @@ from Standard.Base import all
+            duration.hours . should_equal 12
 
         Test.specify "should normalize subtraction" <|
-            interval = 13.months - 1.month
```
**Request** (_at least for explanation_): Removing these tests seems suspicious to me. I believe `month` or `year` is a case that requires `Period` and not `Duration` - does it mean `Period` will no longer be supported, if this PR is integrated? CCing @radeusgd and @jdunkerley .

hubertp replied:
I believe they've been just moved around, right @Akirathan ?

JaroslavTulach reviewed (approved):
The changes that introduce `EnsoDuration` seems fine to me. They directly copy what has been done previously and that is acceptable. Let few notes there related to some future changes...

I have a feeling this PR removes support for `Period`. Not sure whether that is OK or not? (marked as **request**)

Please manually check that native image still builds and can execute something simple. (marked as **request**)

Commit:
Refactor: Use @Builtin.wrapException

Commit:
Fix unused parameter warnings in Date.enso and Time_Of_Day.enso

Commit:
Update CHANGELOG

Commit:
Make all classes implementing TruffleObject final.

Commit:
Fix few tests

JaroslavTulach commented:
Thanks for making the `TruffleObject` implementations `final`, Pavel.

hubertp commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Time_Of_Day.enso:
```diff
@@ -217,8 +217,7 @@ type Time_Of_Day
     + : Duration -> Time_Of_Day
-    + self amount = if amount.is_date then Error.throw (Time_Error_Data "Time_Of_Day does not support date intervals") else
-        Time_Utils.time_adjust self Time_Utils.AdjustOp.PLUS amount.internal_duration
+    + self amount = self.plus_builtin amount
```
Any particular reason why we should prefer builtins rather than calling Java code? This is a general question, also to @JaroslavTulach

JaroslavTulach replied:
Builtins are necessary when we need to deal with `TruffleObject` and use a bit of partial evaluation. Otherwise it is better if the libraries stick with classical (hosted) Java interop.

hubertp commented on test/Tests/src/Data/Time/Date_Spec.enso:
```diff
@@ -396,7 +396,7 @@ js_parse text format=Nothing =
     zone = Time_Zone.utc
     datetime_with_tz = local_datetime.at_zone zone
-    diff = Duration.between datetime_with_tz local_datetime (timezone_aware=False)
+    diff = Duration.between datetime_with_tz local_datetime
```
And this passes now? Call me surprised :)

hubertp commented on test/Tests/src/Semantic/Java_Interop_Spec.enso:
```diff
@@ -1,4 +1,6 @@
 from Standard.Base import all
+import Standard.Test
+import Standard.Base.Data.Time.Duration
```
```suggestion
```
spurious (now)

hubertp commented on test/Tests/src/Semantic/Java_Interop_Spec.enso:
```diff
@@ -53,4 +54,5 @@ spec =
             april1st.month.should_equal 4
             april1st.day.should_equal 1
 
+
```
```suggestion
```

hubertp reviewed (approved):
LGTM. Thanks for making the changes

Commit:
Ignore tests that require new Period type

Commit:
Merge branch 'develop' into wip/akirathan/builtin-duration-type-182987539

Akirathan committed:
Update test/Tests/src/Semantic/Java_Interop_Spec.enso

Co-authored-by: Hubert Plociniczak <hubert.plociniczak@gmail.com>

Akirathan committed:
Update test/Tests/src/Semantic/Java_Interop_Spec.enso

Co-authored-by: Hubert Plociniczak <hubert.plociniczak@gmail.com>

Commit:
Reformat code

Commit:
Fix native image build

Commit:
Add timezone_aware argument to Duration.between

Akirathan committed:
Merge branch 'develop' into wip/akirathan/builtin-duration-type-182987539

Commit:
Run prettier

Commit:
Merge branch 'wip/akirathan/builtin-duration-type-182987539' of github.com:enso-org/enso into wip/akirathan/builtin-duration-type-182987539

radeusgd commented on CHANGELOG.md:
```diff
@@ -209,6 +209,7 @@
 - [Added `Filter_Condition` to `Vector`, `Range` and `List`.][3770]
 - [Extended `Filter_Condition` with `Is_Empty`, `Not_Empty`, `Like` and
   `Not_Like`.][3775]
+- [Reimplemented `Duration` as a built-in type][3759]
```
```suggestion
- [Reimplemented `Duration` as a built-in type.][3759]
```

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Duration.enso:
```diff
@@ -38,25 +54,17 @@ time_execution ~function =
 
 # TODO Dubious constructor export
 from project.Data.Time.Duration.Duration import all
 from project.Data.Time.Duration.Duration export all
```
I think this can be removed no, right?

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Date.enso:
```diff
@@ -5,6 +5,7 @@ import Standard.Base.Data.Time.Date_Period
 
 import Standard.Base.Polyglot
 from Standard.Base.Error.Common import Time_Error_Data
+from Standard.Base.Error.Common import unimplemented
```
```suggestion
from Standard.Base.Error.Common import Time_Error_Data, unimplemented
```

We don't have a style guide and I guess both approaches are OK, I'm wondering if we should keep the imports from a single module in one line? I think most of the time that's what we've been doing.

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDuration.java:
```diff
@@ -0,0 +1,167 @@
+  @Builtin.Method(description = "Gets the nanoseconds")
+  public long nanoseconds() {
+    return duration.toNanos();
+  }
```
I think we should handle/wrap the `ArithmeticException`. Here and also in `milliseconds`.

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Duration.enso:
```diff
@@ -125,14 +158,7 @@ type Duration
-    total_milliseconds self =
-        Panic.catch ArithmeticException (self.internal_duration . toMillis) _->
-            Error.throw (Illegal_Argument_Error_Data "The duration is too large to convert it to milliseconds.")
+    milliseconds self = @Builtin_Method "Duration.milliseconds"
```
If I understand correctly, we change the semantics here.

Earlier this was returning the nanos part of the duration, i.e. if I have a duration of 1s 300ms, it would return `300`, now it will return `1300`. That's why I added the separate `total_milliseconds` method.

Is the semantics change deliberate or accidental?

Still, even if we want to keep the changed semantics, we definitely must update the docs to reflect that - `Get the portion of the duration expressed in milliseconds.` suggests the old semantics!

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Duration.enso:
```diff
@@ -165,61 +191,28 @@ type Duration
-         Convert duration of a year and a hour to a vector returning
-         `[1, 0, 0, 1, 0, 0, 0]`.
+         Convert a duration of one hour to a vector resulting in
+         `[1, 60, 3600, 3600000, 3600000000000]`.
```
I don't understand this change.

Why would I ever need a vector in which the same duration is expressed in various units?

The original meaning of this function was to represent the duration by each part so that the whole vector would reflect the whole duration. Now each element reflects the whole duration (with some partial precision). Do we need such a method?

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Duration.enso:
```diff
@@ -141,9 +167,9 @@ type Duration
-             example_seconds = Examples.duration.milliseconds
+             example_seconds = Examples.duration.seconds
     seconds : Integer
-    seconds self = self.internal_duration . toSecondsPart
+    seconds self = @Builtin_Method "Duration.seconds"
```
Earlier this method would return the seconds _part_ so it made sense for it to be an integer - a part will always be an integer.

Now if we are returning the whole duration in seconds, if we want to keep this new semantics, I think we should reconsider making it return a `Decimal` instead so the millisecond part is not truncated. If it is to be truncated, we must mention in the documentation that the milliseconds get truncated and specify what method of rounding is used.

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDateTime.java:
```diff
@@ -138,6 +144,24 @@ public EnsoTimeZone zone() {
     return new EnsoTimeZone(dateTime.getZone());
   }
 
+  @Builtin.Method(name = "plus_builtin", description = "Adds a duration to this date time")
+  @Builtin.Specialize
+  @Builtin.WrapException(from = UnsupportedMessageException.class, to = PanicException.class)
+  @CompilerDirectives.TruffleBoundary
+  public EnsoDateTime plus(Object durationObject, InteropLibrary interop)
+      throws UnsupportedMessageException {
+    return new EnsoDateTime(dateTime.plus(interop.asDuration(durationObject)));
+  }
```
Normally if a type mismatch happens, we will throw a `Type_Error` or `Unsupported_Argument_Types`.

If I understand this code correctly, here instead we will throw `UnsupportedMessageException`. I think maybe we should wrap it into some Enso-specific type? I think we could re-use one of the two above or alternatively create a new one. I'm just not convinced we should expose a Truffle exception type to our users - instead we should have an exception type that shows up somewhere in Enso like the two above (i.e. it has its `type Type_Error` definition with documentation, `to_display_text` etc.)

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDuration.java:
```diff
@@ -0,0 +1,167 @@
+    } catch (UnsupportedMessageException e) {
+      throw new PanicException(e.getMessage(), interop);
+    }
+  }
```
Do I understand correctly that I can do
`Duration.between (Date.new 1999 1 2) (Date_Time.new 1999 4 8 12 40)` and with `convertToDateTime` the former will be converted to a Date Time @ `1999-01-02 0:00:00` and happily compute the difference?

I think passing a Date/Date_Time pair should result in some Type_Error or Illegal_Argument_Error - we should either have two Dates or two Date_Times, not a mix of them. The mix is going to likely be a user error - if the user really wants to compute a difference between mixed types they can perform a conversion explicitly.

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDuration.java:
```diff
@@ -0,0 +1,167 @@
+    return duration.toSeconds();
+  }
+
+  @Builtin.Method(description = "Gets the milliseconds")
```
This docstring (not sure how important it is and where it shows up?) is a bit unclear. Is this the total duration expressed in milliseconds or the millisecond _part_ of the duration? (I know it is the former after reading the code carefully, but the doc should IMO answer that question without requiring to read the code - otherwise why have the doc at all)

radeusgd commented on test/Tests/src/Data/Time/Duration_Spec.enso:
```diff
@@ -3,99 +3,88 @@ from Standard.Base import all
-            interval.to_json.should_equal <|
-                duration_pairs = [["nanoseconds", interval.nanoseconds], ["seconds", interval.seconds], ["hours", interval.hours], ["days", interval.days], ["months", interval.months], ["years", interval.years]]
-                Json.from_pairs ([["type", "Duration"]] + duration_pairs)
+            # Test units separately, as adding, e.g., seconds modifies also milliseconds
+            (13.nanoseconds.to_json.get "nanoseconds") . to_text . should_equal "13"
```
So where is the test for other units?

radeusgd reviewed (commented):
The code looks good, but as mentioned in the in-line comments, the semantics of the `seconds`, `milliseconds` etc. getters has been altered.

I want to confirm if this change was deliberate, as I'm not aware of any discussions to change this part of the design. IMO the old way where you could get each part was useful - and we could also have the `total_` variants of each to get the full duration expressed in a given unit - giving us both capabilities.

Still, if it was deliberate and we want to keep the new semantics, the documentation needs to be updated, and `milliseconds` and `nanoseconds` getters should handle the `ArithmeticException`, converting it to some dataflow error (maybe `Illegal_State_Error`?).

And we should revise if we want the `to_vector` method on Duration - with the changed semantics it does not seem useful at all.

Commit:
Revert changes in the Duration API

Commit:
Remove Duration Integer extensions.

Specify the Duration rather with explicit constructor.

Commit:
Fix after merge

Commit:
Builtins return long instead of int

Akirathan commented:
As discussed with @radeusgd, I reverted the incompatible changes that I introduced in this PR. So the only breaking changes are:
* removing all `Integer` extension methods, as we want to explicitly differentiate between `Duration` and `Period`.
* Ignore all tests that are related to `Period`, i.e., that will be failing untill `Period` is implemented.

Commit:
More test fixes

Akirathan committed:
Apply suggestions from code review

Co-authored-by: Radosław Waśko <radoslaw.wasko@enso.org>

Commit:
Date_Time.plus/minus and Date.plus/minus handles ArithmeticException

Commit:
Duration.between accepts only Date_Time

Commit:
Fix native image build

Commit:
Merge branch 'develop' into wip/akirathan/builtin-duration-type-182987539

mergify[bot] committed:
Merge branch 'develop' into wip/akirathan/builtin-duration-type-182987539

Commit:
Merge branch 'wip/akirathan/builtin-duration-type-182987539' of github.com:enso-org/enso into wip/akirathan/builtin-duration-type-182987539

Commit:
Fix Http_Spec

mergify[bot] committed:
Merge branch 'develop' into wip/akirathan/builtin-duration-type-182987539

mergify[bot] committed:
Merge branch 'develop' into wip/akirathan/builtin-duration-type-182987539

mergify[bot] committed:
Merge branch 'develop' into wip/akirathan/builtin-duration-type-182987539

mergify[bot] committed:
Merge branch 'develop' into wip/akirathan/builtin-duration-type-182987539

mergify[bot] committed:
Merge branch 'develop' into wip/akirathan/builtin-duration-type-182987539

mergify[bot] committed:
Merge branch 'develop' into wip/akirathan/builtin-duration-type-182987539

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Date_Time.enso:
```diff
@@ -387,9 +388,10 @@ type Date_Time
+             example_plus = Date_Time.new 2020 + 15.years + (Duration.new hours=3)
+    + : Duration -> Date_Time ! Polyglot_Error
+    + self amount =
+        Panic.catch ArithmeticException (self.plus_builtin amount) (err -> Error.throw (Polyglot_Error_Data err.getMessage))
```
I don't think this should be wrapped in `Polyglot_Error_Data`. This is a wrapper in which Java exceptions are usually wrapped - so the field `cause` is supposed to be the exception, not a `Text` message. And we have never (IIRC) used `Polyglot_Error` as a dataflow error - it's a marker for Java exceptions coming from builtin/polyglot calls _and_ there are plans to remove it too. So I'm against using it in this context.

Please use some more fitting exception. Maybe `Illegal_State_Error`? Or maybe `Time_Error` - I think it is fitting what we need here.

radeusgd commented on distribution/lib/Standard/Base/0.0.0-dev/src/Data/Time/Duration.enso:
```diff
@@ -191,24 +192,46 @@ type Duration
+    total_milliseconds : Integer ! Illegal_Argument_Error
+    total_milliseconds self =
+        Panic.catch ArithmeticException (self.total_milliseconds_builtin) _->
+            Error.throw (Illegal_Argument_Error_Data "The duration is too large to convert it to milliseconds")
```
minor nitpick: this function has no arguments (okay, other than `self`), so maybe `Illegal_State_Error`? Because the result is independent of arguments - it depends on the internal _state_ of `self`.

But feel free to leave as-is too, I don't think that's very important. Maybe we should decide on some robust guideline when to use which.

radeusgd commented on distribution/lib/Standard/Test/0.0.0-dev/src/Main.enso:
```diff
@@ -166,7 +166,7 @@ specify : Text -> Any -> (Text | Nothing) -> Nothing
     pair = case pending of
         Nothing -> Duration.time_execution (run_spec behavior)
-        reason -> Pair_Data 0.milliseconds (Pending reason)
+        reason -> Pair_Data Duration.new (Pending reason)
```
I'm wondering if it wouldn't be more readable if we made a `Duration.zero` alias which calls `new` with no arguments. Would be clearer that it is a 0-duration.

Commit:
Get rid of unpacked build dependencies.

See https://www.pivotaltracker.com/story/show/183543026

radeusgd reviewed (commented):
Please change the `Polyglot_Error` before merging, otherwise gtg.

radeusgd commented on engine/runtime/src/main/java/org/enso/interpreter/runtime/data/EnsoDuration.java:
```diff
@@ -0,0 +1,195 @@
+        return LocalDateTime.of(date, time);
+      }
+    } catch (UnsupportedMessageException e) {
+      throw new PanicException(e.getMessage(), interop);
```
This is an internal error that should never happen, right? Because we just assert that `isDate && isTime`.

Then I think the convention is to throw `IllegalStateException`, see for example `ContextRewrapNode.java:113`.

And other than that for future - the I think `throw new PanicException(x)` essentially acts as `Panic.throw x` in Enso.
So what you are doing is `Panic.throw e.getMessage` - i.e. throwing a pure `Text` value. We actually used to do that a lot, but have since moved to throwing meaningful atoms - because if you throw just Text it's hard to catch it - you can just `Panic.catch Text` but you cannot do more filtering easily, whereas we have facilities for catching a specific exception type. So in general it would be good to put some atom as the payload for such panics (the atom itself may for example contain a varying message, but it should be some type so that we can distinguish types of exceptions).

Commit:
Add Duration.zero - alias for Duration.new

Commit:
Use more appropriate errors

Commit:
Merge branch 'develop' into wip/akirathan/builtin-duration-type-182987539
